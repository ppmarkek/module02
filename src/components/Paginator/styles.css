.paginator {
  display: flex;
  gap: 5px;
}

.paginatorButton {
  all: unset;
  width: 30px;
  height: 30px;
  border: 1px solid #8083a3;
  border-radius: 5px;
  cursor: pointer;
  transition: all 0.5s;
  display: flex;
  align-items: center;
  justify-content: center;
}

.paginatorButton:hover {
  border: 1px solid #6b59cc;
}

.activePaginatorButton {
  all: unset;
  width: 30px;
  height: 30px;
  border: 1px solid #6b59cc;
  background-color: #6b59cc;
  color: #fff;
  border-radius: 5px;
  cursor: pointer;
  transition: all 0.5s;
  display: flex;
  align-items: center;
  justify-content: center;
}

.activePaginatorButton:hover {
  border: 1px solid #8083a3;
  background-color: #fff;
  color: #000;
}

.paginatorWrapper {
  height: 50px;
  width: 100%;
  display: flex;
  gap: 10px;
  flex-direction: column;
  align-items: center;
}

.paginationButton {
  all: unset;
  width: 30px;
  height: 30px;
  border: 1px solid #8083a3;
  border-radius: 5px;
  cursor: pointer;
  transition: all 0.5s;
  display: flex;
  align-items: center;
  justify-content: center;
}

.paginationButton:hover {
  border: 1px solid #6b59cc;
}
