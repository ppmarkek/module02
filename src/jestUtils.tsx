import React from 'react';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import { reducer } from './redux/index';
import { createStore } from '@reduxjs/toolkit';

export const renderWithRedux = (ui: React.JSX.Element, state = {}) => {
  const Wrapper = ({
    children,
  }: {
    children: React.ReactNode | React.ReactNode[];
  }) => (
    <Router>
      <Provider store={createStore(reducer, state)}>{children}</Provider>
    </Router>
  );

  return render(ui, { wrapper: Wrapper });
};
