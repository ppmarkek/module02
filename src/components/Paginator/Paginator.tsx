import { useSearchParams } from 'react-router-dom';
import './styles.css';
import { useGetAllRequestedResultsQuery } from '../../redux/apiService';
import { useSearch } from '../../useSearch';

type PaginatorProps = {
  currentPage?: number;
  totalPages?: number;
};

export const Paginator = ({ currentPage, totalPages }: PaginatorProps) => {
  const [searchParams, setSearchParams] = useSearchParams();
  const { searchResult, requestResults, isLoading } = useSearch();

  useGetAllRequestedResultsQuery(
    {
      searchCriteria: searchResult,
      size: requestResults.itemsPerPage,
      page: currentPage?.toString() ?? '1',
    },
    {
      skip: !currentPage,
    }
  );

  const selectPage = (page: number) => {
    setSearchParams({
      ...Object.fromEntries(searchParams),
      page: page.toString(),
    });
  };

  return (
    <div className="paginatorWrapper">
      {!isLoading && (
        <div className="paginator" data-testid="paginator">
          <button
            onClick={() => selectPage(currentPage! - 1)}
            data-testid="page-button-previous"
            className="paginationButton"
            disabled={currentPage! <= 0}
          >
            &lt;
          </button>
          {[...Array(totalPages).keys()].map((page, index) => (
            <div
              key={index}
              data-testid={`page-button-${index}`}
              className={
                index === currentPage
                  ? 'activePaginatorButton'
                  : 'paginatorButton'
              }
              onClick={() => selectPage(index)}
            >
              {page + 1}
            </div>
          ))}
          <button
            onClick={() => selectPage(currentPage! + 1)}
            data-testid="page-button-next"
            className="paginationButton"
            disabled={currentPage! >= totalPages! - 1}
          >
            &gt;
          </button>
        </div>
      )}
    </div>
  );
};
