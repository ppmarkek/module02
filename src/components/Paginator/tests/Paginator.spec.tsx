import { renderWithRedux } from '../../../jestUtils';
import { appStoreMock } from '../../../fixtures/results';
import '@testing-library/jest-dom';
import { screen } from '@testing-library/react';
import { Paginator } from '../Paginator';

jest.mock('../../../redux/apiService', () => {
  const mockData = {
    results: [
      { id: 1, title: 'Test Item 1' },
      { id: 2, title: 'Test Item 2' },
    ],
  };

  const mockCategoriesData = {
    categories: ['Category 1', 'Category 2', 'Category 3'],
  };
  const originalModule = jest.requireActual('../../../redux/apiService');

  const mockUseGetAllRequestedResultsQuery = jest.fn().mockReturnValue({
    data: mockData,
    refetch: jest.fn().mockResolvedValue({ data: mockData }),
  });

  const mockUseGetCategoriesQuery = jest.fn().mockReturnValue({
    data: mockCategoriesData,
    refetch: jest.fn().mockResolvedValue({ data: mockCategoriesData }),
  });

  return {
    ...originalModule,
    useGetAllRequestedResultsQuery: mockUseGetAllRequestedResultsQuery,
    useGetCategoriesQuery: mockUseGetCategoriesQuery,
  };
});

jest.mock('next/navigation', () => ({
  ...require('next-router-mock'),
  useSearchParams: () => jest.fn(),
}));

jest.mock('next/router', () => ({
  useRouter() {
    return {
      route: '/',
      pathname: '',
      query: '',
      asPath: '',
      push: jest.fn(),
      events: {
        on: jest.fn(),
        off: jest.fn(),
      },
      beforePopState: jest.fn(() => null),
      prefetch: jest.fn(() => null),
    };
  },
}));

describe('Paginator component', () => {
  it('should render correctly', () => {
    const component = renderWithRedux(<Paginator totalPages={5} />, {
      appStore: appStoreMock,
    });

    expect(component).toMatchSnapshot();
  });

  it('should not render component while loading', () => {
    renderWithRedux(<Paginator totalPages={5} />);

    expect(screen.queryAllByTestId('paginator')).toHaveLength(1);
  });

  it('should component back button be disabled while on first page', () => {
    const { getByTestId } = renderWithRedux(
      <Paginator totalPages={5} currentPage={0} />
    );

    expect(
      getByTestId('prev-pagination-button').closest('button')
    ).toHaveAttribute('disabled');
  });

  it('should component next button be disabled while on last page', () => {
    const { getByTestId } = renderWithRedux(
      <Paginator totalPages={5} currentPage={4} />
    );

    expect(
      getByTestId('next-pagination-button').closest('button')
    ).toHaveAttribute('disabled');
  });
});
