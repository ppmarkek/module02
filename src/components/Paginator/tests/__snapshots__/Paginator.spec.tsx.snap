// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paginator component should render correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="paginatorWrapper"
      >
        <div
          class="paginator"
          data-testid="paginator"
        >
          <button
            class="paginationButton"
            data-testid="first-pagination-button"
          >
            &lt;
          </button>
          <div
            class="paginatorButton"
          >
            1
          </div>
          <div
            class="paginatorButton"
          >
            2
          </div>
          <div
            class="paginatorButton"
          >
            3
          </div>
          <div
            class="paginatorButton"
          >
            4
          </div>
          <div
            class="paginatorButton"
          >
            5
          </div>
          <button
            class="paginationButton"
            data-testid="second-pagination-button"
          >
            &gt;
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="paginatorWrapper"
    >
      <div
        class="paginator"
        data-testid="paginator"
      >
        <button
          class="paginationButton"
          data-testid="first-pagination-button"
        >
          &lt;
        </button>
        <div
          class="paginatorButton"
        >
          1
        </div>
        <div
          class="paginatorButton"
        >
          2
        </div>
        <div
          class="paginatorButton"
        >
          3
        </div>
        <div
          class="paginatorButton"
        >
          4
        </div>
        <div
          class="paginatorButton"
        >
          5
        </div>
        <button
          class="paginationButton"
          data-testid="second-pagination-button"
        >
          &gt;
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
